react-native-maps/react-native-maps
# Typescript rewrite

### Does any other open PR do the same thing?

No

### What issue is this PR fixing?

None

### How did you test this PR?

Source code evaluation and manual testing.

### Misc comments

A full transition to TS.

This took a quite a while to do, so reviewing probably will as well.

This fixes numerous (I stopped counting) TS errors, mostly in callback types, and probably also introduces some (I did rigorous examination of every component, but I surely missed something), and as such, this _might_ introduce new errors.
Regardless, this should lead the way for a new code base which doesn't require manual examination of the d.ts file and proptypes each time the source code is changed.

We could also look into generating docs from JSDoc, to further reduce the sources of documentation? I have no personal experience with this, but reducing the different parts of documentation that needs manual work would be great.

The documentation currently uses `@platform iOS` and `@platform Android` without further description, but as iOS supports both Google Maps and Apple Maps, this isn't enough. This pull request seeks to address this as well, adding a thorough platform description to each component prop.

Please feel free to chime in regarding type placement and so on. This probably needs some refactoring.

I still need to do some testing on this as well, but the more the merrier, so I decided to push this now.

Testing:
1. Clone this pull request
2. Run the newly added build script, eg: yarn build
3. copy and replace dist/index.js, dist/index.d.ts, dist/types.js, dist/types.d.ts & dist/components folder into your node_modules/react-native-maps

BREAKING CHANGES:
1. drops supports for react-native < 0.63.4
2. Remove animateToNavigation, animateToCoordinate, animateToViewingAngle and animateToBearing.
3. Removes legacy iOS snapshot method
4. Components are no longer re-exported under MapView.
